File: api/forecast.ts
import type { VercelRequest, VercelResponse } from '@vercel/node';

// Proxies Open-Meteo 7-day forecast (no API key required)
// Query params: lat, lon

export default async function handler(req: VercelRequest, res: VercelResponse) {
  const lat = req.query.lat || '39.5';
  const lon = req.query.lon || '-98.35';

  const url = `https://api.open-meteo.com/v1/forecast?latitude=${lat}&longitude=${lon}&daily=temperature_2m_max,temperature_2m_min,precipitation_sum,precipitation_probability_max,wind_speed_10m_max,weather_code&temperature_unit=fahrenheit&wind_speed_unit=mph&precipitation_unit=inch&timezone=auto&forecast_days=7`;

  try {
    const response = await fetch(url);
    const data = await response.json();
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Cache-Control', 's-maxage=3600, stale-while-revalidate=600');
    res.json(data);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch forecast' });
  }
}

File: backend/ingest/forecast.py
"""
Open-Meteo 7-day forecast ingestion.
No API key required.
"""
import logging
import requests

logger = logging.getLogger(__name__)

OPEN_METEO_URL = "https://api.open-meteo.com/v1/forecast"


def fetch_forecast(lat: float, lon: float) -> dict:
    """
    Fetch 7-day daily forecast from Open-Meteo.
    Returns the raw JSON response shaped for the frontend ForecastPanel.
    """
    params = {
        "latitude": lat,
        "longitude": lon,
        "daily": [
            "temperature_2m_max",
            "temperature_2m_min",
            "precipitation_probability_max",
            "weathercode",
            "windspeed_10m_max",
            "precipitation_sum",
        ],
        "temperature_unit": "fahrenheit",
        "wind_speed_unit": "mph",
        "precipitation_unit": "inch",
        "forecast_days": 7,
        "timezone": "auto",
    }

    try:
        resp = requests.get(OPEN_METEO_URL, params=params, timeout=10)
        resp.raise_for_status()
        data = resp.json()
        return data
    except requests.RequestException as e:
        logger.error(f"Open-Meteo fetch error: {e}")
        return {
            "error": str(e),
            "daily": {
                "time": [],
                "temperature_2m_max": [],
                "temperature_2m_min": [],
                "precipitation_probability_max": [],
                "weathercode": [],
                "windspeed_10m_max": [],
                "precipitation_sum": [],
            },
        }
